File: csharp/SimpleHash.cs
using System;

// Token: 0x020001DD RID: 477
public struct SimpleHash
{
	// Token: 0x060015D1 RID: 5585 RVA: 0x0017BD77 File Offset: 0x00179F77
	public void SetEmpty()
	{
		this.bits = 0U;
	}

	// Token: 0x060015D2 RID: 5586 RVA: 0x0017BD80 File Offset: 0x00179F80
	public void InitHashBits(float x, float y, float z)
	{
		int num = (int)((x + 200f) / 40f);
		int num2 = (int)((y + 200f) / 40f);
		int num3 = (int)((z + 200f) / 40f);
		num = ((num < 9) ? ((num < 0) ? 0 : num) : 9);
		num2 = ((num2 < 9) ? ((num2 < 0) ? 0 : num2) : 9);
		num3 = ((num3 < 9) ? ((num3 < 0) ? 0 : num3) : 9);
		this.bits = (1U << num | 1024U << num2 | 1048576U << num3);
	}

	// Token: 0x060015D3 RID: 5587 RVA: 0x0017BE12 File Offset: 0x0017A012
	public bool MaskPass(uint mask)
	{
		return (mask & this.bits) == this.bits;
	}

	// Token: 0x060015D4 RID: 5588 RVA: 0x0017BE24 File Offset: 0x0017A024
	public static uint GenerateHashMask(float x, float y, float z, int area = 1)
	{
		int num = (int)((x + 200f) / 40f);
		int num2 = (int)((y + 200f) / 40f);
		int num3 = (int)((z + 200f) / 40f);
		num = ((num < 9) ? ((num < 0) ? 0 : num) : 9);
		num2 = ((num2 < 9) ? ((num2 < 0) ? 0 : num2) : 9);
		num3 = ((num3 < 9) ? ((num3 < 0) ? 0 : num3) : 9);
		uint num4 = 0U;
		num4 |= 1U << num;
		num4 |= 1024U << num2;
		num4 |= 1048576U << num3;
		if (area > 9)
		{
			area = 9;
		}
		for (int i = 1; i <= area; i++)
		{
			if (num >= i)
			{
				num4 |= 1U << num - i;
			}
			if (num2 >= i)
			{
				num4 |= 1024U << num2 - i;
			}
			if (num3 >= i)
			{
				num4 |= 1048576U << num3 - i;
			}
			if (num < 10 - i)
			{
				num4 |= 1U << num + i;
			}
			if (num2 < 10 - i)
			{
				num4 |= 1024U << num2 + i;
			}
			if (num3 < 10 - i)
			{
				num4 |= 1048576U << num3 + i;
			}
		}
		return num4;
	}

	// Token: 0x04001C27 RID: 7207
	public uint bits;
}

File: src/data_struct/simple_hash.rs
pub struct SimpleHash {
    bits: u32,
}

impl SimpleHash {
    pub fn set_empty(&mut self) {
        self.bits = 0;
    }

    pub fn init_hash_bits(&mut self, x: f32, y: f32, z: f32) {
        let mut x_index = ((x + 200.0) / 40.0) as i32;
        let mut y_index = ((y + 200.0) / 40.0) as i32;
        let mut z_index = ((z + 200.0) / 40.0) as i32;
        x_index = if x_index < 9 {
            if x_index < 0 {
                0
            } else {
                x_index
            }
        } else {
            9
        };
        y_index = if y_index < 9 {
            if y_index < 0 {
                0
            } else {
                y_index
            }
        } else {
            9
        };
        z_index = if z_index < 9 {
            if z_index < 0 {
                0
            } else {
                z_index
            }
        } else {
            9
        };
        self.bits = (1 << x_index | 1024 << y_index | 1048576 << z_index);
    }

    pub fn mask_pass(&self, mask: u32) -> bool {
        (mask & self.bits) == self.bits
    }

    pub fn generate_hash_mask(x: f32, y: f32, z: f32, mut area: i32) -> u32 {
        let mut x_index = ((x + 200.0) / 40.0) as i32;
        let mut y_index = ((y + 200.0) / 40.0) as i32;
        let mut z_index = ((z + 200.0) / 40.0) as i32;
        x_index = if x_index < 9 {
            if x_index < 0 {
                0
            } else {
                x_index
            }
        } else {
            9
        };
        y_index = if y_index < 9 {
            if y_index < 0 {
                0
            } else {
                y_index
            }
        } else {
            9
        };
        z_index = if z_index < 9 {
            if z_index < 0 {
                0
            } else {
                z_index
            }
        } else {
            9
        };
        let mut hash_mask = 0;
        hash_mask |= 1 << x_index;
        hash_mask |= 1024 << y_index;
        hash_mask |= 1048576 << z_index;
        if area > 9 {
            area = 9;
        }
        for i in 1..=area {
            if x_index >= i {
                hash_mask |= 1 << x_index - i;
            }
            if y_index >= i {
                hash_mask |= 1024 << y_index - i;
            }
            if z_index >= i {
                hash_mask |= 1048576 << z_index - i;
            }
            if x_index < 10 - i {
                hash_mask |= 1 << x_index + i;
            }
            if y_index < 10 - i {
                hash_mask |= 1024 << y_index + i;
            }
            if z_index < 10 - i {
                hash_mask |= 1048576 << z_index + i;
            }
        }
        hash_mask
    }
}
